src/core/exec-invoke.c

```diff
@@ -2440,6 +2448,91 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
+        else if ((size_t) r >= buffer_size) {
+                log_debug("Exec quotas: Failed to generate project id for %s", target_dir);
+                return -1;
+        }
+
+        uint32_t proj_id = MurmurHash2(proj_id_plain, strlen(proj_id_plain), 0);
+
+        /* Check if project quotas are supported */
+        struct dqblk req;
+        r = quotactl_devnum(QCMD_FIXED(Q_GETQUOTA, PRJQUOTA), devno, proj_id, &req);
+        if (r == -ESRCH || ERRNO_IS_NEG_NOT_SUPPORTED(r)) {
+                log_warning("Not applying storage quotas: project quotas are not supported for %s", target_dir);
+                return 0;
+        } else if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to query disk quota for prjid %u: %m", proj_id);
+
+        /* Set projid (equivalent to chattr +P -p) */
+        struct fsxattr attrs;
+        int fd = xopenat(AT_FDCWD, target_dir, O_RDONLY|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW);
+        r = ioctl(fd, FS_IOC_FSGETXATTR, &attrs);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to get fsxattr for %s", target_dir);
+
+        attrs.fsx_xflags = FS_XFLAG_PROJINHERIT;
+        attrs.fsx_projid = proj_id;
+
+        r = ioctl(fd, FS_IOC_FSSETXATTR, &attrs);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to set project id for %s: %m", target_dir);
+
+        /* Set Quotas */
+        _cleanup_free_ char *devpath = NULL;
+        r = device_open_from_devnum(S_IFBLK, devno, O_RDONLY|O_CLOEXEC, &devpath);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to get device path for %s: %m", target_dir);
+
+        _cleanup_free_ char *devmountpoint = NULL;
+        r = find_mountpoint(devpath, &devmountpoint);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to find device mountpoint for %s: %m", target_dir);
+
+        struct statvfs disk_st;
+        r = statvfs(devmountpoint, &disk_st);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to get filesystem data for %s: %m", target_dir);
+
+        double percent = UINT32_SCALE_TO_PERCENT(percent_quota) / 100.0;
+        uint64_t block_limit = (uint64_t)(disk_st.f_blocks * percent);
+        uint64_t inode_limit = (uint64_t)(disk_st.f_files * percent);
+
+        req.dqb_bhardlimit = block_limit;
+        req.dqb_bsoftlimit = 0;
+        req.dqb_ihardlimit = inode_limit;
+        req.dqb_isoftlimit = 0;
+        req.dqb_valid = QIF_LIMITS;
+        r = quotactl_devnum(QCMD_FIXED(Q_SETQUOTA, PRJQUOTA), devno, proj_id, &req);
+        if (r < 0)
+                return log_debug_errno(r, "Exec quotas: Failed to set project quotas for %s: %m", target_dir);
+
+        log_debug("Storage quotas set for %s. Block limit = %lu, inode limit = %lu", target_dir, block_limit, inode_limit);
+
+        return r;
```

Comment:
by switching this to fd-based operation things would automatically start working with tmpfs as backend btw, i.e. also with RuntimeDirectory